### PR TITLE
[WIP] bazel, tensorflow: enable Darwin source build

### PR DESCRIPTION
**Pull request by lukeadams**
###### Motivation for this change
Tracking for WIP patches.

 * Need to break apart commits since some Bazel commits are mixed in with the tf commit
 * need to set LD to re-enable bazel tests on Darwin
 * also need to make the changes Darwin only since linux doesn't need these hacks. usually just lib.optionalString
###### Things done
With this PR tf is buildable for py27/py36, sidestepping the unicode mismatch when using a wheel for py27. This also enables tf c++ artifacts, but I haven't tested those yet.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/tools/build-managers/bazel/default.nix [hunk 1]
```diff
@@ -71,7 +92,7 @@ stdenv.mkDerivation rec {
 
   # Build the CPP and Java examples to verify that Bazel works.
 
-  doCheck = true;
+  doCheck = false;
```
Comment (matthewbauer) [inline pkgs/development/tools/build-managers/bazel/default.nix]: This should be a conditional, I think.

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 2]
```diff
@@ -105,11 +109,32 @@ let
...
+        export LD="$(pwd)/ld.sh"
+
+        echo "#!/usr/bin/env bash" >> test.sh
+        chmod +x test.sh
+        declare -px >> test.sh
+        cp test.sh cc.sh
+        cp test.sh cxx.sh
+        cp test.sh ld.sh
+
+        echo "(clang \"\$@\" ) || (clang++ \"\$@\" )" >> cc.sh
```
Comment (matthewbauer) [inline pkgs/development/python-modules/tensorflow/default.nix]: Have you tried just doing "clang++" for this? I think that clang++ should support both (although some of the Nix purity stuff may break this).

**Comment by matthewbauer** [comment]
Thanks for working on this! Tensorflow on Darwin would be great.

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 3]
```diff
@@ -88,7 +92,7 @@ let
 
     hardeningDisable = [ "all" ];
 
-    bazelFlags = [ "--config=opt" ]
+    bazelFlags = [ "--config=opt" "--config=monolithic" ]
```
Comment (abbradar) [inline pkgs/development/python-modules/tensorflow/default.nix]: Please leave a comment on why is this necessary; let's make it optional for Darwin to avoid static dependencies as much as possible.

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 4]
```diff
@@ -58,7 +61,8 @@ let
     nativeBuildInputs = [ swig which ];
 
     buildInputs = [ python jemalloc openmpi glibcLocales numpy ]
-      ++ lib.optionals cudaSupport [ cudatoolkit cudnn nvidia_x11 ];
+      ++ lib.optionals cudaSupport [ cudatoolkit cudnn nvidia_x11 ]
```
Comment (abbradar) [inline pkgs/development/python-modules/tensorflow/default.nix]: Mark `stdenv.isDarwin && cudaSupport` as broken.

File: pkgs/development/python-modules/tensorflow/default.nix (hunk 2 above)
Comment (lukeadams) [reply pkgs/development/python-modules/tensorflow/default.nix]: I initially tried this, but it caused failures when building a few dependencies. I'll try to find a log so I can at least put a comment explaining it.

File: pkgs/development/tools/build-managers/bazel/default.nix (hunk 1 above)
Comment (lukeadams) [reply pkgs/development/tools/build-managers/bazel/default.nix]: I'll change this to `(!stdenv.isDarwin)` if I can't get tests working

**Comment by LnL7** [comment]
@GrahamcOfBorg build bazel

File: pkgs/development/python-modules/tensorflow/default.nix (hunk 4 above)
Comment (lukeadams) [reply pkgs/development/python-modules/tensorflow/default.nix]: 👍

**Comment by lukeadams** [comment]
@abbradar I cleaned it up quite a bit and added conditionals on everything! There are still a few whitespaces changes I need to make though. Rebased onto master, and am currently testing the py27 build (py36 succeeded).

File: pkgs/development/python-modules/tensorflow/default.nix (hunk 4 above)
Comment (lukeadams) [reply pkgs/development/python-modules/tensorflow/default.nix]: Would it be a good idea to remove the asserts at the top? (see https://github.com/NixOS/nixpkgs/issues/36229)

**Comment by LnL7** [comment]
@GrahamcOfBorg build bazel

**Comment by LnL7** [comment]
Do you have the CLT/Xcode installed?

**Comment by lukeadams** [comment]
@LnL7 Ahh yep (just CLT) – good catch. Since I need it on my main machine, I can work it out Friday on a VM without it installed.

Sorry about that!

**Comment by LnL7** [comment]
That might explain the missing framework, if you add it I can start another build.

**Comment by lukeadams** [comment]
@abbradar I'm not going to have time to finish this for the next month, so maybe you can merge the wheel as python3 only? 

@LnL7 I added the framework, but there are other issues due to the way Bazel calls xcrun/xcbuild. It may be simple to patch, but I don't have time to play with it

**Comment by LnL7** [comment]
@GrahamcOfBorg build bazzel

**Comment by lukeadams** [comment]
@LnL7 The framework was in my staging branch; I just pushed it to this one

**Comment by LnL7** [comment]
@GrahamcOfBorg build bazel

**Comment by abbradar** [comment]
@lukeadams No sweat! I've merged https://github.com/NixOS/nixpkgs/pull/37044 for now.

**Comment by LnL7** [comment]
I was hoping the framework was the only issue, but I get an `xcrun` failure locally. 😕

**Comment by lukeadams** [comment]
Closed until I have time to work on this. Not sure when tbh :/

#40424 with `full-paths.patch` patch looks interesting for Darwin
